proposals/LP0004-project-governance.md

```diff
@@ -0,0 +1,393 @@
+
+In dividing the LLVM project into _area teams_, this proposal seeks to break up areas of the project based on contribution volume and number of active contributors. This proposal assumes that areas of the project with more unique contributors are more likely to have contributor disagreements, and that areas with more highly active contributors will be more likely to have diverse _area team_ formations.
+
+This proposal puts a possible set of _area teams_ to form with the specified areas of ownership. One notable point is that LLVM (`:/llvm`) had 945 unique authors contribute over the 12 month period used for statistics. This is far more than any other part of the project, and it justified splitting LLVM into two teams, which are separated as the _LLVM Backend_ and _LLVM Middle-end_. Clang had 706 unique authors, which is also significantly higher than other sub-projects. To represent different parts of the Clang community the _Clang_ and _Clang Tooling_ teams are separated below.
+
+* LLVM Backend (CodeGen, MC, Target)
+* LLVM Middle-end (Everything not covered by LLVM Backend)
+* Clang (All the `clang` sub-project except tooling and format libraries and tools)
+* Clang Tooling (`clang-tools-extra` and the tooling and format libraries from `clang`)
+* C/C++ Runtime Libraries (`libcxx`, `libc`, `libunwind`, `libcxxabi`)
+* Compiler Runtimes (`compiler-rt`, `openmp`)
+* Flang (`flang`)
+* MLIR (`mlir`)
+* Binary tools (`lldb`, `lld`, `bolt`)
+* Incubator (`circt`, `torch-mlir`, `Polygeist`, `clangir`, etc)
+* Project Infrastructure (`zorg`, `utils`, `llvm-test-suite`, CI/CD, hosting, etc)
+* Community (Discord, Discourse, etc)
+
+##### Community & Infrastructure Area Teams
+
+The _Community_ and _Infrastructure_ _area teams_ have different responsibilities from the other _area teams_ because they aren’t restricted to specific software components of the LLVM project. Their goals are more holistic to provide for the needs of the whole project and wider community.
+
+Both the _community_ and _infrastructure area teams_ will select one member of their team to act as a liaison with the LLVM Foundation. The liaisons will be responsible for managing the relationship between their _area team_ and the employees and directors of the LLVM Foundation.
+
+##### Voting by Area
+
+Every _voting contributor_ is eligible to vote for the project-wide _Community_ and _Infrastructure_ teams. Eligibility for voting on other _area teams_ is determined by _meaningful contributions_ to the areas those teams cover. For example, in order to vote on the **LLVM Backend** _area team_, a contributor would need to contribute code to the LLVM `CodeGen`, `MC`, or `Target` libraries, or to vote on the members of the **Flang** team, a contributor would need to contribute code to the `flang` project. There is no limit to the number of _area teams_ a contributor can vote on if they have made contributions to the relevant areas.
```

Comment:
"would need to contribute code" seems too strong given we explicitly call out things like issue triaging and code review as meaningful contributions for being a voting contributor?

Reply:
We could allow people who don't qualify automatically to apply for voting access to a particular team and then have that team vote on it.  That would help relieve some of the burden of trying to come up with a perfect metric.

What other metrics could we look at besides number of commits that are easy to quantify?

Reply:
I don't know what the metric should be, I'm just saying the document isn't self-consistent at the moment, since we reuse the term "meaningful contribution" here but then go on to say it's only code contributions

Reply:
e.g. if someone's code reviews for a specific backend automatically qualify them for project-wide votes then they should also automatically qualify them for LLVM Backend votes

Reply:
I'm working on some updates, which will make the document consistent. One of the downsides of a document that went through multiple revisions is sometimes vestiges of the old stick around a bit too long 😞.

I was playing around with GitHub's REST API to see what data I could get out of GitHub, and I'm going to spend some time poking at [CHAOSS](https://chaoss.community/). The talk on CHAOSS at the LLVM Dev Meeting was really neat and they had a lot of metrics gathering tools which we may be able to leverage.

Reply:
I'm starting to think that Voting by Area will be too complicated.  It means we need to maintain 12 separate voting lists, 12 separate voting criteria, etc.  How big of a problem would it be if everyone was eligible to vote for every area team?

Reply:
> This seems to fly in the face of _"We value and recognize the many shapes that contributions take."_

I do want to acknowledge this comment. You're totally right. I'm not happy with this. I've struggled to find good ways to support voting by area that didn't rely on commit filtering.

> I'm starting to think that Voting by Area will be too complicated. It means we need to maintain 12 separate voting lists, 12 separate voting criteria, etc. How big of a problem would it be if everyone was eligible to vote for every area team?

Maybe this is the answer. I didn't originally have voting by area. Maybe the right solution here is to make it clear that it is okay for people to abstain from voting in areas that they're not familiar with. I personally can't imagine voting for Flang or MLIR's representation since I know so few of the people working in those areas.

Reply:
Yup. To complement my point on the other thread, artificial limitations are always biased and disproportionately affect certain sub-groups. However, _natural_ limitations are much better, especially when we assume that the people vote when they have a stake or knowledge on the area, not because they're in some sort of a list. We'll get less votes, but more relevant percentage of votes.

And if we notice people trying to put out irrelevant votes just for the sake of being annoying, the CoC can deal with that already.

Reply:
IIUC, the plan is to mostly exercise voting when electing various members/community representatives. Do we plan to use such voting mechanisms to resolve disputes/disagreements?

Reply:
No, the intent is not to enable direct democratic decision making.